File: src/python/aiassisted/downloader.py
"""Download utilities for fetching files from GitHub"""

import tempfile
from pathlib import Path
from typing import Optional
import httpx
from rich.console import Console

console = Console()


class Downloader:
    """Handle file downloads with error handling"""

    def __init__(self, verbosity: int = 1):
        self.verbosity = verbosity

    def log_debug(self, message: str):
        """Log debug message"""
        if self.verbosity >= 2:
            console.print(f"[blue][DEBUG][/blue] {message}")

    def download_file(self, url: str, output_path: Path) -> bool:
        """Download file from URL to output path

        Args:
            url: URL to download from
            output_path: Path to save file to

        Returns:
            True on success, False on failure
        """
        self.log_debug(f"Downloading {url} to {output_path}")

        try:
            with httpx.Client(follow_redirects=True, timeout=30.0) as client:
                response = client.get(url)
                response.raise_for_status()

                output_path.parent.mkdir(parents=True, exist_ok=True)
                output_path.write_bytes(response.content)

                return True
        except httpx.HTTPError as e:
            console.print(
                f"[red][ERROR][/red] Failed to download {url}: {e}", file=sys.stderr
            )
            return False
        except Exception as e:
            console.print(
                f"[red][ERROR][/red] Unexpected error downloading {url}: {e}",
                file=sys.stderr,
            )
            return False

    def download_text(self, url: str) -> Optional[str]:
        """Download text content from URL

        Args:
            url: URL to download from

        Returns:
            Text content on success, None on failure
        """
        self.log_debug(f"Downloading {url}")

        try:
            with httpx.Client(follow_redirects=True, timeout=30.0) as client:
                response = client.get(url)
                response.raise_for_status()
                return response.text
        except httpx.HTTPError as e:
            console.print(
                f"[red][ERROR][/red] Failed to download {url}: {e}", file=sys.stderr
            )
            return None
        except Exception as e:
            console.print(
                f"[red][ERROR][/red] Unexpected error downloading {url}: {e}",
                file=sys.stderr,
            )
            return None


import sys

File: src/bun/src/downloader.ts
/**
 * Download utilities for fetching files from GitHub
 */

import { writeFile, mkdir } from 'fs/promises';
import { dirname } from 'path';

export interface DownloaderOptions {
  verbosity?: number;
}

export class Downloader {
  private verbosity: number;

  constructor(options: DownloaderOptions = {}) {
    this.verbosity = options.verbosity ?? 1;
  }

  private logDebug(message: string): void {
    if (this.verbosity >= 2) {
      console.log(`\x1b[34m[DEBUG]\x1b[0m ${message}`);
    }
  }

  async downloadFile(url: string, outputPath: string): Promise<boolean> {
    this.logDebug(`Downloading ${url} to ${outputPath}`);

    try {
      const response = await fetch(url);
      
      if (!response.ok) {
        console.error(`\x1b[31m[ERROR]\x1b[0m Failed to download ${url}: HTTP ${response.status}`);
        return false;
      }

      const content = await response.arrayBuffer();
      
      // Ensure parent directory exists
      await mkdir(dirname(outputPath), { recursive: true });
      
      // Write file
      await writeFile(outputPath, new Uint8Array(content));
      
      return true;
    } catch (error) {
      console.error(`\x1b[31m[ERROR]\x1b[0m Failed to download ${url}: ${error}`);
      return false;
    }
  }

  async downloadText(url: string): Promise<string | null> {
    this.logDebug(`Downloading ${url}`);

    try {
      const response = await fetch(url);
      
      if (!response.ok) {
        console.error(`\x1b[31m[ERROR]\x1b[0m Failed to download ${url}: HTTP ${response.status}`);
        return null;
      }

      return await response.text();
    } catch (error) {
      console.error(`\x1b[31m[ERROR]\x1b[0m Failed to download ${url}: ${error}`);
      return null;
    }
  }
}
